### DISCUSSION TITLE
Recommended practices for tools to listen to nREPL evaluations

### DISCUSSION THREAD

**Discussion by daslu** [nrepl/nrepl#277] [General]
Some tools like to use nREPL middleware to listen to the user's code evaluations.

Examples:
* [Reveal](https://vlaaad.github.io/reveal/) can listen to evaluations and send values to be visualized.
* The recent [Portal-Clerk-Kindly-nREPL experiment](https://github.com/scicloj/visual-tools-experiments/tree/main/portal-clerk-kindly-nrepl-1) does similarly with [Portal](https://github.com/djblue/portal).
* [Notespace](https://github.com/scicloj/notespace/) is attempting to listen to all user evaluations in the editor & REPL and automatically maintain a live-reloading document that makes sense out of them (still incomplete, on hold at the moment). 

Making it work requires some setup on the user side.
The required setup depends on the choice of tooling (clj, lein) x (CIDER, Calva, ... ) x (Reveal, Notespace, ...).

Here is, for example, [How to use Calva and Reveal together](https://calva.io/reveal/) (through either clj or lein). It is quite nicely explained, but still could be troublesome for a new user.

I am wondering about recommended practices for making such situations seamless to the user.

In particular, is it somehow possible for tools & libraries to listen to code evaluations without defining their own middleware (and needing the user to add the middleware)?
To make that happen, probably they should have some way to recognize the correct nREPL server and hook to it.

**Comment by daslu** [comment]
As a concrete example, see the current Setup instructions of the Clay tool -- the `Setup` subsection in the intro: 
https://scicloj.github.io/clay/#/notebooks/intro.clj

For CIDER, the situation seems a bit more confusing than for Calva -- one needs to write a [.dir-locals.el](https://www.gnu.org/software/emacs/manual/html_node/emacs/Directory-Variables.html) file, realize that Emacs is not always refreshing its knowledge after a change in that file, etc.
When discussing this with users, this seems like a current point of friction in getting things right.

**Comment by daslu** [comment]
As commented to me by @generateme, the emacs variable `cider-jack-in-nrepl-middewares` is useful for injecting middleware. E.g.:
```clj 
((clojure-mode . ((eval . (add-to-list 'cider-jack-in-nrepl-middlewares "scicloj.clay.v1.nrepl/middleware")))))
```
I will adapt the Clay intro and examples to recommend that.

**Comment by bbatsov** [comment]
Btw, if someone is going to be using `clay` all the time they can also add something like this directly to their `init.el` or equivalent. (together with the dependency, of course) And for Lein users the setup process can be simplified if there was a Lein plugin that does the dependency and middleware injection itself (that's how `cider-nrepl`, `refactor-nrepl` and `sayid`) do things. See https://github.com/clojure-emacs/sayid/blob/master/src/sayid/plugin.clj for an example.
